langpuppet/merge.go

```diff
@@ -0,0 +1,165 @@
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/pgraph"
+)
+
+// This method returns the merged graph containing all vertices
+// and edges found in the graphs produced by the lang and Puppet GAPIs
+// associated with the wrapping GAPI.
+// Vertices are merged if they adhere to the following rules (for any
+// given value of POSTFIX):
+//  1. The graph from lang contains a noop vertex named puppet_POSTFIX
+//  2. The graph from Puppet contains an empty class mgmt_POSTFIX
+//  3. The resulting graph will contain one noop vertex named POSTFIX
+//     that replaces all nodes mentioned in (1) and (2)
+// All edges connecting to any of the vertices merged this way
+// will be present in the merged graph.
+func mergeGraphs(graphFromLang, graphFromPuppet *pgraph.Graph) (*pgraph.Graph, error) {
+	if graphFromLang == nil || graphFromPuppet == nil {
+		return nil, fmt.Errorf("cannot merge graphs until both child graphs are loaded")
+	}
+
+	result, err := pgraph.NewGraph(graphFromLang.Name + "+" + graphFromPuppet.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	mergeTargets := make(map[string]pgraph.Vertex)
+
+	// first add all vertices from the lang graph
+	for _, vertex := range graphFromLang.Vertices() {
+		if strings.Index(vertex.String(), "noop[puppet_") == 0 {
+			resource, ok := vertex.(engine.Res)
+			if !ok {
+				return nil, fmt.Errorf("vertex %s is not a named resource", vertex.String())
+			}
+			basename := strings.TrimPrefix(resource.Name(), "puppet_")
+			// FIXME: will this change the original vertex?
+			resource.SetName(basename)
+			mergeTargets[basename] = vertex
+		}
+		result.AddVertex(vertex)
+		for _, neighbor := range graphFromLang.OutgoingGraphVertices(vertex) {
+			result.AddVertex(neighbor)
+			result.AddEdge(vertex, neighbor, graphFromLang.FindEdge(vertex, neighbor))
+		}
+	}
+
+	var anchor pgraph.Vertex
+	mergePairs := make(map[pgraph.Vertex]pgraph.Vertex)
+
+	// do a scan through the Puppet graph, and mark all vertices that will be
+	// subject to a merge, so it will be easier do generate the new edges
+	// in the final pass
+	for _, vertex := range graphFromPuppet.Vertices() {
+		if vertex.String() == "noop[admissible_Stage[main]]" {
+			// we can start a depth first search here
+			anchor = vertex
+			continue
+		}
+		// at this stage we don't distinguis between class start and end
+		if strings.Index(vertex.String(), "noop[admissible_Class[Mgmt_") != 0 &&
+			strings.Index(vertex.String(), "noop[completed_Class[Mgmt_") != 0 {
+			continue
+		}
+
+		resource, ok := vertex.(engine.Res)
+		if !ok {
+			return nil, fmt.Errorf("vertex %s is not a named resource", vertex.String())
+		}
+		// strip either prefix (plus the closing bracket)
+		basename := strings.TrimSuffix(
+			strings.TrimPrefix(
+				strings.TrimPrefix(resource.Name(),
+					"admissible_Class[Mgmt_"),
+				"completed_Class[Mgmt_"),
+			"]")
+
+		if _, found := mergeTargets[basename]; !found {
+			// FIXME: should be a warning not an error?
```

Comment:
That question can always be answered by a flag if you want. But I think error. Really it's up to you.